core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -144,288 +141,188 @@ public final BasicInvocationStage visitClearCommand(InvocationContext ctx, Clear
-         return CompletableFuture.completedFuture(value);
-
-      final CacheTopology cacheTopology = stateTransferManager.getCacheTopology();
-      final int currentTopologyId = cacheTopology.getTopologyId();
+   protected final CompletableFuture<InternalCacheEntry> retrieveFromProperSource(
+         Object key, AbstractTopologyAffectedCommand command, boolean isWrite) throws Exception {
+      final CacheTopology cacheTopology = checkTopologyId(command);
 
-      if (trace) {
-            log.tracef("Perform remote get for key %s. topologyId=%s, currentTopologyId=%s",
-                       key, lastTopologyId, currentTopologyId);
-      }
-      List<Address> targets;
-      int newTopologyId;
-      if (lastTopologyId < currentTopologyId) {
-         // Cache topology has changed or it is the first time.
-         newTopologyId = currentTopologyId;
-         ConsistentHash readCH = cacheTopology.getReadConsistentHash();
-         boolean isLocal;
-         List<Address> owners;
-         if (readCH.isReplicated()) {
-            isLocal = readCH.isSegmentLocalToNode(rpcManager.getAddress(), 0);
-            owners = readCH.getMembers();
-         } else {
-            owners = readCH.locateOwners(key);
-            isLocal = owners.contains(rpcManager.getAddress());
-         }
-         if (isLocal) {
-            // Normally looking the value up in the local data container doesn't help, because we already
-            // tried to read it in the EntryWrappingInterceptor.
-            // But ifIf we became an owner in the read CH after EntryWrappingInterceptor, we may not find the
-            // value on the remote nodes (e.g. because the local node is now the only owner).
-            return CompletableFuture.completedFuture(dataContainer.get(key));
-         } else {
-            targets = new ArrayList<>(owners);
-         }
-      } else if (lastTopologyId == currentTopologyId && cacheTopology.getPendingCH() != null) {
-         // Same topologyId, but the owners could have already installed the next topology
-         // Lets try with pending consistent owners (the read owners in the next topology)
-         newTopologyId = currentTopologyId + 1;
-         targets = new ArrayList<>(cacheTopology.getPendingCH().locateOwners(key));
-         // Remove already contacted nodes
-         targets.removeAll(cacheTopology.getReadConsistentHash().locateOwners(key));
-         if (targets.isEmpty()) {
+      ConsistentHash readCH = cacheTopology.getReadConsistentHash();
+      List<Address> owners;
+      if (readCH.isReplicated()) {
+         if (readCH.isSegmentLocalToNode(rpcManager.getAddress(), 0)) {
             if (trace) {
-               log.tracef("No valid values found for key '%s' (topologyId=%s).", key, currentTopologyId);
+               log.tracef("Key %s is local, skipping remote get", key);
             }
             return CompletableFutures.completedNull();
          }
+         owners = readCH.getMembers();
```

Comment:
first, shouldn't we try to fetch the most up-to-date value?
second, isn't the goal of the messy staggered gets to fetch the most up-to-date value from the primary before trying the backups?